modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -136,10 +137,27 @@
 
       (isa? coercion :Coercion/YYYYMMDDHHMMSSString->Temporal)
       {"$dateFromString" {:dateString field-name
-                          :format "%Y%m%d%H%M%S"
-                          :onError field-name}}
+                          :format     "%Y%m%d%H%M%S"
+                          :onError    field-name}}
 
-      :else field-name)))
+      ;; mongo only supports datetime
+      (isa? coercion :Coercion/ISO8601->DateTime)
+      {"$dateFromString" {:dateString field-name
+                          :onError    field-name}}
+
+
+      (isa? coercion :Coercion/ISO8601->Date)
+      (throw (ex-info (tru "MongoDB does not support parsing strings as dates. Try parsing to a datetime instead")
+                      {:type              qp.error-type/unsupported-feature
+                       :coercion-strategy coercion}))
```

Comment:
What do think about composing our own parser for dates with MongoDB expressions? e.g.

```suggestion
      (isa? coercion :Coercion/ISO8601->DateTime)
      {"$dateFromString" {:dateString field-name
                          :onError    field-name}}

      (isa? coercion :Coercion/ISO8601->Date)
      {"$dateFromString" {:dateString {"$substr" [field-name 0 10]}
                          :onError    field-name}}
```

Reply:
I think we shouldn't support `Coercion/ISO8601->Date` because even if we could parse the `datestring` it's still a `datetime` type in mongoDb.

In all of our SQL drivers when `Coercion/ISO8601->Date` we convert it to a `date` column and we should keep the behavior consistence even that means we don't support for Mongo.

Reply:
I understood the purpose of coercion strategies is to be able to convert between base types and effective types. Effective types don't need to be represented in the DB.

See the docstring for the `metabase.types` namespace:
https://github.com/metabase/metabase/blob/aa30353a164a6febb6c932ba65208cf023d62eba/shared/src/metabase/types.cljc#L20

The example given in that docstring is for a DB that doesn't support `:type/DateTime`. It explains how we can still support `:Coercion/ISO8601->DateTime` with the following types:
- base_type: :type/Text
- effective_type: :type/DateTime
- a coercion strategy of :Coercion/ISO8601->DateTime

Reply:
I worry that if we add support for `date` and our users might run into correctness problems because `2000/01/01 00:00:00` is not exactly = `2000/01/01`.

I can't think of an example right now but I'd imagine solving those problems in Mongo would not be easy, 
The worse part is we'll have to solve it because we already offered the feature.

So I'd rather avoid putting us in that position.

Reply:
Makes sense, I also haven't fully resolved the implications of doing this.

My thinking was that `2000/01/01 00:00:00` is isomorphic to `2000/01/01` in every case except formatting on the frontend, where the effective_type is used. That assumption could easily be wrong. Let's leave it for now.